contributors/design-proposals/scheduling/scheduling-policy.md

```diff
@@ -0,0 +1,516 @@
+
+```
+
+### policy composition
+
+`bindingMode` Describes how policies should be composed:
+
+- any: any policy that has its rules satisfied
+- all: all policies MUST have its rules satisfied
+
+if we have a heterogenous bindingMode across policies (i.e. some policies with any and others with all).
+Then the most restrictive one (i.e the all bindingMode) is applied.
+
+### unset fields:
+
+if a field was not specified in the policy no rule should apply (i.e. everything is allowed).
+
+### empty match:
+
+Usually policies should distinguish between empty matches, since it depends on the action:
+
+- require/deny: no rule apply.
+- allow: allows everything.
+
+
+### inside the matches:
+
+The match field can express pretty much any structure you want to, but there's some things that should be considered:
+
+- the structure should match as much as possible what you try to take action on.
+- match elements are ANDed.
+- the `*` wildcard is usually needed in the allow section. It should be represented with an empty value (e.g. empty array).
+
+
+### conflict handling for policies
+
+There is two kinds of conflict in policies:
+
+- structural conflicts: these exists when any rule has the same matchingExpression and opposed actions (deny and require)
+
+- semantical conflicts: these exists due to semantic of the fields (e.g. require a NodeSelector `master=true` and also require `master=false`)
+
+
+Structural conflicts must handled at creation time, on the other hand, semantical conflicts should be handled at runtime: detect that there's a conflict and emit an event stating that policies couldn't be satisfied due to a conflict.
+
+
+### required vs allowed vs denied
+
+Policies may have overlapping rules, to handle this policies are computed in the following order:
+
+- compute what is denied.
+- compute what is required.
+- compute what was allowed.
+
+they should also obey to the following rules:
+
+- everything that is required is by definition allowed.
+- everything that is not denied is not automatically allowed.
+
+
+# Detailed Design
+
+
+## SchedulingPolicy
+
+Proposed API group: `policy/v1alpha1`
+
+
+### SchedulingPolicy content
+
+SchedulingPolicy spec is composed of optional fields that allow scheduling rules. If a field is absent from a SchedulingPolicy, this `SchedulingPolicy` won't allow any item from the missing field.
+
+```yaml
+apiVersion: policy/v1alpha1
+kind: SchedulingPolicy
+metadata:
+  name: mySchedulingPolicy
+spec:
+  bindingMode: all
+  namespaces:
+    - default
+  rules:
+    schedulerNames:
+      - match: []
+        action:
+    priorityClassNames:
+      - match: []
+        action:
+    tolerations:
+      - match: []
+        action:
+    nodeSelectors:
+      - match: []
+        action:
+    nodeAffinities:
+      - match: []
+        action:
+    podAntiAffinities:
+      - match: []
+        action:
+    podAffinities:
```

Comment:
Not sure about podAffinities. What kind of use-cases do you have in mind for pod affinities in scheduling policy.

Reply:
the use case would be a cache trashing application. you want to enforce the fact that it specifies a podAffinity with other cache trashing applications to co-localize them.

Reply:
In the case of cache trashing application, I think it makes more sense to limit them to a set of nodes, instead of forcing them to specify podAffinity towards one another.